flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java

```diff
@@ -103,11 +105,16 @@ protected void serializeOperatorState(OperatorState operatorState, DataOutputStr
         serializeStreamStateHandle(operatorState.getCoordinatorState(), dos);
 
         // Sub task states
-        final Map<Integer, OperatorSubtaskState> subtaskStateMap = operatorState.getSubtaskStates();
-        dos.writeInt(subtaskStateMap.size());
-        for (Map.Entry<Integer, OperatorSubtaskState> entry : subtaskStateMap.entrySet()) {
-            dos.writeInt(entry.getKey());
-            serializeSubtaskState(entry.getValue(), dos);
+        if (operatorState.isFullyFinished()) {
+            dos.writeInt(-1);
```

Comment:
nit: to me, explicitly writing `isFullyFinished` would be more clear and less error-prone:
```
dos.writeBoolean(operatorState.isFullyFinished());
dos.writeInt(subtaskStateMap.size()); // zero
// empty loop
```

Reply:
I'm a bit hesitate here since we should not be able to add new field to the meta data since it would impact the compatibility. There would be the cases that user taking a savepoint or external checkpoint with an older version (like 1.12) and resume the savepoint/checkpoint with new version (like the 1.13) to do the upgrade. If we add new fields, we should not e able to parse the existing savepoint / checkpoint ?

Reply:
I see @gaoyunhaii concerns. I'd be fine with using `-1` here.